#Dropout Regularization Technique In Deep Neural Network

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import tensorflow as tf
from tensorflow import keras
%matplotlib inline
import seaborn as sns

In [8]:
path = '/content/sonar_dataset.csv'
df = pd.read_csv(path, header= None)

In [9]:
df.head()

0       1       2       3       4   ...      56      57      58      59  60
0  0.0200  0.0371  0.0428  0.0207  0.0954  ...  0.0180  0.0084  0.0090  0.0032   R
1  0.0453  0.0523  0.0843  0.0689  0.1183  ...  0.0140  0.0049  0.0052  0.0044   R
2  0.0262  0.0582  0.1099  0.1083  0.0974  ...  0.0316  0.0164  0.0095  0.0078   R
3  0.0100  0.0171  0.0623  0.0205  0.0205  ...  0.0050  0.0044  0.0040  0.0117   R
4  0.0762  0.0666  0.0481  0.0394  0.0590  ...  0.0072  0.0048  0.0107  0.0094   R

[5 rows x 61 columns]

In [10]:
df.shape

(208, 61)

In [11]:
#Checking NaN Values
df.isna().sum()

0     0
1     0
2     0
3     0
4     0
     ..
56    0
57    0
58    0
59    0
60    0
Length: 61, dtype: int64

In [12]:
df.columns

Int64Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
            17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
            34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
            51, 52, 53, 54, 55, 56, 57, 58, 59, 60],
           dtype='int64')

In [16]:
features = df.drop(60, axis = 'columns')
target = df[60]

In [17]:
target

0      R
1      R
2      R
3      R
4      R
      ..
203    M
204    M
205    M
206    M
207    M
Name: 60, Length: 208, dtype: object

In [19]:
target = pd.get_dummies(target, drop_first = True)
target.sample(5)

R
91   1
109  0
143  0
40   1
85   1

M becomes 0 and R = 1

In [21]:
target.value_counts()

R
0    111
1     97
dtype: int64

M = 111, R = 97

In [22]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(features, target, test_size=0.25, random_state=1)

In [23]:
X_train.head()

0       1       2       3   ...      56      57      58      59
67   0.0368  0.0403  0.0317  0.0293  ...  0.0135  0.0067  0.0078  0.0068
14   0.0124  0.0433  0.0604  0.0449  ...  0.0114  0.0196  0.0147  0.0062
164  0.0163  0.0198  0.0202  0.0386  ...  0.0084  0.0108  0.0083  0.0033
179  0.0394  0.0420  0.0446  0.0551  ...  0.0101  0.0068  0.0053  0.0087
19   0.0126  0.0149  0.0641  0.1732  ...  0.0181  0.0094  0.0116  0.0063

[5 rows x 60 columns]

In [24]:
model = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dense(1, activation='sigmoid')
])

model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

model.fit(X_train, y_train, epochs=100, batch_size=8)

Epoch 1/100
20/20 [==============================] - 2s 3ms/step - loss: 0.6900 - accuracy: 0.5513
Epoch 2/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6658 - accuracy: 0.6090
Epoch 3/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6417 - accuracy: 0.6603
Epoch 4/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6151 - accuracy: 0.6923
Epoch 5/100
20/20 [==============================] - 0s 2ms/step - loss: 0.5877 - accuracy: 0.7244
Epoch 6/100
20/20 [==============================] - 0s 2ms/step - loss: 0.5643 - accuracy: 0.6987
Epoch 7/100
20/20 [==============================] - 0s 2ms/step - loss: 0.5142 - accuracy: 0.7436
Epoch 8/100
20/20 [==============================] - 0s 3ms/step - loss: 0.4807 - accuracy: 0.8205
Epoch 9/100
20/20 [==============================] - 0s 2ms/step - loss: 0.4481 - accuracy: 0.7885
Epoch 10/100
20/20 [==============================] - 0s 2ms/step - loss: 0.4214 - accuracy: 0.8333
Epoch 11/

In [25]:
model.evaluate(X_test, y_test)

2/2 [==============================] - 0s 8ms/step - loss: 1.1837 - accuracy: 0.7500


[1.1837193965911865, 0.75]

In [26]:
y_pred = model.predict(X_test).reshape(-1)
print(y_pred[:10])

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

2/2 [==============================] - 0s 5ms/step
[1.1096997e-08 9.9872488e-01 9.9027044e-01 6.2783424e-06 9.9999982e-01
 9.8523277e-01 7.1417377e-03 1.0000000e+00 1.9025545e-05 1.0000000e+00]
[0. 1. 1. 0. 1. 1. 0. 1. 0. 1.]


In [27]:
from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.71      0.89      0.79        27
           1       0.83      0.60      0.70        25

    accuracy                           0.75        52
   macro avg       0.77      0.74      0.74        52
weighted avg       0.77      0.75      0.74        52



##Now, Applying Dropout method instead of doing flat-wise calculation in the Neural Networks

In [29]:
model_dropout = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(1, activation='sigmoid')
])

model_dropout.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

model_dropout.fit(X_train, y_train, epochs=100, batch_size=8)

Epoch 1/100
20/20 [==============================] - 1s 4ms/step - loss: 0.6979 - accuracy: 0.5128
Epoch 2/100
20/20 [==============================] - 0s 3ms/step - loss: 0.7137 - accuracy: 0.4744
Epoch 3/100
20/20 [==============================] - 0s 4ms/step - loss: 0.7262 - accuracy: 0.4744
Epoch 4/100
20/20 [==============================] - 0s 3ms/step - loss: 0.7004 - accuracy: 0.4808
Epoch 5/100
20/20 [==============================] - 0s 3ms/step - loss: 0.7080 - accuracy: 0.4808
Epoch 6/100
20/20 [==============================] - 0s 4ms/step - loss: 0.6871 - accuracy: 0.5000
Epoch 7/100
20/20 [==============================] - 0s 4ms/step - loss: 0.6963 - accuracy: 0.5513
Epoch 8/100
20/20 [==============================] - 0s 3ms/step - loss: 0.7061 - accuracy: 0.5000
Epoch 9/100
20/20 [==============================] - 0s 3ms/step - loss: 0.6877 - accuracy: 0.5321
Epoch 10/100
20/20 [==============================] - 0s 4ms/step - loss: 0.6849 - accuracy: 0.5128
Epoch 11/

In [30]:
model_dropout.evaluate(X_test, y_test)

2/2 [==============================] - 0s 14ms/step - loss: 0.4195 - accuracy: 0.7885


[0.41948360204696655, 0.7884615659713745]

In [31]:
y_pred = model_dropout.predict(X_test).reshape(-1)
print(y_pred[:10])

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

2/2 [==============================] - 0s 5ms/step
[0.00206235 0.7752739  0.9321529  0.0412912  0.9967831  0.8428884
 0.49367458 0.9980517  0.06852008 0.9984546 ]
[0. 1. 1. 0. 1. 1. 0. 1. 0. 1.]


In [32]:
from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.75      0.89      0.81        27
           1       0.85      0.68      0.76        25

    accuracy                           0.79        52
   macro avg       0.80      0.78      0.78        52
weighted avg       0.80      0.79      0.79        52



## **Result**: test accuracy increased from 0.77 to 0.81 by using Dropout Regularization. However, dropout doesn't always suit the model.